DESIGN.md

```diff
@@ -0,0 +1,279 @@
+of what bits serve what purpose.
+
+## Identity
+
+Fleetspeak identifies clients using a fingerprint of a public key, much like
+openssl keys. Specifically, client creates a keypair on startup, or when asked
+to rekey. The Fleetspeak
+[common.ClientID](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/common#ClientID)
+the first 8 bytes of a sha256 hash of the public part of this key. It is the
+[Communicator's](#server-communicator) responsibility to verify the identity of
+clients. The recommended approach is to communicate over TLS using the client's
+Fleetspeak key as TLS client identification. The
+[https.Communicator source](https://github.com/google/fleetspeak/blob/master/fleetspeak/src/server/https/https.go)
+has an example of this.
+
+## Messages
+
+At its core, Fleetspeak forwards
+[fleetspeak.Message](https://github.com/google/fleetspeak/blob/master/fleetspeak/src/common/proto/fleetspeak/common.proto)
+protocol buffers. The comments in the `.proto` file should be considered
+definitive, but the most important fields to understand are `destination`,
+`message_type` and `data`. Essentially, `destination` field tells Fleetspeak
+where to forward the message to. The `message_type` and `data` fields are
+forwarded verbatim, and when integrating a service a developer should choose
+these values according to their needs.
+
+The `message_type` is meant to be used both for dispatching within a service,
+and for reporting. So for example if your integration has three basic types of
+messages to pass, you might indicate them with three different message type
+strings. By doing this you make it possible to query Fleetspeak for the number
+of messages of each type.
+
+# Server
+
+A Fleetspeak Server middleware process written in go. To start a server, you
+instantiate a
+[server.Server](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/server#Server),
+providing the components needed for your particular installation.
+
+
+## Miniserver: An Example
+The
+[miniserver](https://github.com/google/fleetspeak/blob/master/fleetspeak/src/demo/miniserver/miniserver.go)
+binary from the demo directory defines a Fleetspeak Server in a way suitable for
+demonstrations and small installations.
+
+When run, the miniserver binary binds two ports, and accepts bind addresses for
+each from the command line. The `--https_addr` flag determines the interface and
+port that clients will connect to. It needs to be open to the internet, or at
+least to all Fleetspeak client machines. See [Server
+Communicator](#server-communicator), below.
+
+The `--admin_addr` flag determines the interface and port that the
+[Administrative Interface](#administrative-interface) listens on. The miniserver
+binary does not perform any authentication of administrative requests. Therefore
+access to this port needs to be limited to trusted processes.
+
+The miniserver process stores all of its state in an sqlite version 3
```

Comment:
Nit: s/sqlite/Sqlite/

Reply:
Actually, according to official website and Wikipedia, it should be *SQLite*.

Reply:
s/sqlite/SQLite/ done here and a few other places.

Reply:
Nit, opt.: s/Sqlite/SQLite/ still needed.

Reply:
Done.